### DISCUSSION TITLE
Timestamps in GPX File

### DISCUSSION THREAD

**Discussion by dcwuser** [jaybo/deepzoom-blog#46] [Q&A] [answered]
I am trying to prepare a GPX file describing a (multi-route, multi-day, marine) trip to others. I would like to send them the file and have them be able to see start times and end times (and ideally also waypoint times) along the route each day. The GPX XML schema supports setting a time for each waypoint (https://www.topografix.com/GPX/1/1/#type_wptType). But when I use `<time>` elements to specify timestamps and import the GPX file, these times do not appear in the Depart and Arrive UI elements in either the Trips or Routes tabs. And when I set these times in the UI and then export the trip to GPX, the GPX contains no `<time>` elements. Is there a way to get DeepZoom to do what I want or does DeepZoom just not deal with GPX timestamps?

**Comment by jaybo** [comment]
Here's the current situation:
### GPX Routes
1. The GPX XML schema for routes doesn't have a standardized representation for speed and departure time.  These _could_ be inferred from `<time>` fields on each `<rtept>` but this is not currently implemented.  At present, on import only `opencpn:planned_departure` and `opencpn:planned_speed` are supported.  On export, departure time and speed are not implemented, nor is a `<time>` property added to `<rtept>`.

### GPX Tracks
1. On import after conversion to geojson if `feature.properties.coordinateProperties.times` exists, it will be retained and used to timestamp each coordinate.  Attached is an example created by RideWithGPS of my favorite bike ride around Seattle (with stops at multiple coffee shops).  You can then convert the track to a route.
[SeattleBikeLoop.zip](https://github.com/user-attachments/files/17806638/SeattleBikeLoop.zip)

So for your purpose, it sounds like you're using some other program to generate the .gpx file, and mainly want the `<time>` fields to be retained on import, yes?  Aside:  What program are you using and could you please provide a sample .gpx file?

Frankly, when working on this I got so disheartened after understanding the chaos of conflicting .gpx representations that I deprioritized the effort.   But it sounds like I should just do the `<time>` property inference on export and import.

**Reply by dcwuser** [reply]
Thanks for your comprehensive reply, and for your work on DeepZoom. I looked at the opencpn specification https://mail.opencpn.org/wiki/dokuwiki/doku.php?id=opencpn:manual_basic:route_manager:use_gpx_files and added `<opencpn:planned_speed>` and `<opencpn:planned_departure>` elements to my first `<rte>`, but when I import into DeepZoom it they still don't seem to show up on the Routes, Trips, or animation. I have attached my GPX file (with extension changed to TXT so github will allow it), which I have written by hand. Can you tell me what I am doing wrong?

[DavidAltRoute.txt](https://github.com/user-attachments/files/17808266/DavidAltRoute.txt)

For my purposes, supporting `<time>` on `<rtept>` would indeed be sufficient, but if industry-standard tools are producing these other tags, that might be a better approach.

**Reply by jaybo** [reply]
Regarding DavidAltRoute.txt, it also needs to have `creator="OpenCPN"`, as in:

`<gpx version="1.1" creator="OpenCPN" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://www.topografix.com/GPX/1/1" xmlns:gpxx="http://www.garmin.com/xmlschemas/GpxExtensions/v3" xsi:schemaLocation="http://www.topografix.com/GPX/1/1 http://www.topografix.com/GPX/1/1/gpx.xsd http://www.garmin.com/xmlschemas/GpxExtensions/v3 http://www8.garmin.com/xmlschemas/GpxExtensionsv3.xsd" xmlns:opencpn="http://www.opencpn.org">
`

**Reply by dcwuser** [reply]
Thanks again! I've become a paid subscriber. I see that the creator and version attributes are indeed required by the schema, and when I add creator="OpenCPN" I do get the starting time from the `<opencpn:planned_departure>` attribute of the first `<rte>`. The corrected file is attached.

[DavidAltRoute2.txt](https://github.com/user-attachments/files/17810268/DavidAltRoute2.txt)

However...

1. The planned_departure time of the 1st route is used for all the routes. Is this a bug or am I still doing it wrong?
2. (Less important) It assumes the timestamp is UTC, even if it doesn't have the final Z. According to ISO8601, timestamps with no UTC relation given should be assumed to be local. I can fix this, though, by appending a -08:00 offset.
3. (Less important) It appears to work only if I specify creator="OpenCPN". I don't think this is how creator attribute is supposed to work; requiring this specific creator string is essentially saying that not only did OpenCPN introduce these tags, but no other tool is allowed to use them (or has to lie and say it is OpenCPN in order to be allowed to).

While I think it's nice to support the OpenCPN tags, from my perspective, supporting timestamps specified on each rtept would indeed be preferable. That would also let me specify a pause for a given time, or different speeds along different parts of a route. 

Thanks again for your help.

**Reply by jaybo** [reply]
1. Yes, a bug.  Working on it.
2. I need to look at this...
3. I agree, and will remove the `creator` requirement.

The big problem is that none of the available gpx/kml parsers handle metadata generically (I'm using https://www.npmjs.com/package/@tmcw/togeojson) .  Even though they perform the task of conversion to geojson, I'll need to end up writing my own parallel parser just for the metadata.  **Argh**!

**Comment by jaybo** [comment]
Update:  I have import working for **DavidAltRoute2.txt**.  I'm still working on export, so please don't test this quite yet.
1. Time and speed of each route correctly imported.
2. Local time should be accepted along with UTC.
3. Removed `creator` requirement.

Please use https://deepzoom-test.azurewebsites.net/ for testing, as this is not officially released yet.

Here's the related pull request in **togeojson**.  https://github.com/placemark/togeojson/pull/129

**Comment by jaybo** [comment]
Import and Export of Markers, Tracks, and Routes round trip through GPX completed at  https://deepzoom-test.azurewebsites.net/ .
This update will be in the next release.

**Comment by jaybo** [comment, marked as answer]
This is now fixed in release V3.15.3
